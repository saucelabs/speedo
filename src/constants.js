export const USAGE = `
Speedo CLI runner`

export const EPILOG = 'Copyright 2019 Â© Sauce Labs'

export const PERFORMANCE_METRICS = [
    'timeToFirstByte',
    'firstPaint',
    'firstContentfulPaint',
    'firstMeaningfulPaint',
    'domContentLoaded',
    'timeToFirstInteractive',
    'load',
    'speedIndex',
    'perceptualSpeedIndex',
    'pageWeight',
    'pageWeightEncoded'
]

const METRIC_PARAM = {
    alias: 'm',
    name: 'metric',
    description: 'metric that you want to check (multiple possible)',
    default: 'load'
}

const ALL_PARAM = {
    name: 'all',
    description: 'check on all metrics',
    default: false,
    type: 'boolean'
}

export const COMMON_CLI_PARAMS = [{
    alias: 'h',
    name: 'help',
    description: 'prints speedo help menu'
}, {
    alias: 'u',
    name: 'user',
    description: 'your Sauce Labs username'
}, {
    alias: 'k',
    name: 'key',
    description: 'your Sauce Labs user key'
}, {
    alias: 'r',
    name: 'region',
    description: 'your Sauce Labs datacenter region, the following regions are available: `us-west-1` (short `us`), `eu-central-1` (short `eu`)',
    default: 'us'
}, {
    alias: 'l',
    name: 'logDir',
    description: 'directory to store logs from testrun'
}, {
    alias: 't',
    name: 'traceLogs',
    description: 'if set runner downloads tracing logs for further investigations'
}]

export const RUN_CLI_PARAMS ={
    platformName: {
        alias: 'p',
        description: 'the platform the performance test should run in (e.g. "Windows 10")',
        default: 'Windows 10'
    },
    browserVersion: {
        alias: 'v',
        description: 'the browser version of Chrome the performance test should run in (e.g. "latest")',
        default: 'latest'
    },
    build: {
        alias: 'b',
        description: 'name of the build you want to run your performance test in'
    },
    name: {
        alias: 'n',
        description: 'name of your performance test'
    },
    metric: METRIC_PARAM,
    all: ALL_PARAM,
    retry: {
        description: 'amount of retries for failing performance tests',
        default: 0,
        type: 'number'
    },
    tunnelIdentifier: {
        alias: 'i',
        description: 'identifier for Sauce Connect tunnel to run performance tests for local hosted apps'
    }
}

export const ANALYZE_CLI_PARAMS = {
    name: {
        alias: 'n',
        description: 'name of a specific test within given build to analyze'
    },
    orderIndex: {
        alias: 'o',
        description: 'number of page you have opened in that test (requires to specify a test name with -n)'
    },
    pageUrl: {
        alias: 'u',
        description: 'url of page in the test you want to analyze (requires to specify a test name with -n)'
    },
    metric: METRIC_PARAM,
    all: ALL_PARAM
}

export const ERROR_MISSING_CREDENTIALS = `
Your Sauce credentials are missing!
Either set 'SAUCE_USERNAME' and 'SAUCE_ACCESS_KEY' in your environment or
provide them as parameter`

export const REQUIRED_TESTS_FOR_BASELINE_COUNT = 10
export const JOB_COMPLETED_TIMEOUT = 20000
export const JOB_COMPLETED_INTERVAL = 1000
