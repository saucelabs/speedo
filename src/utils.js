import chalk from 'chalk'
import { table } from 'table'
import prettyMs from 'pretty-ms'
import prettyBytes from 'pretty-bytes'

import { JOB_COMPLETED_TIMEOUT, JOB_COMPLETED_INTERVAL, PERFORMANCE_METRICS } from './constants'

/**
 * print results of cli run
 * @param  {Object}   result            result of performance assertion
 * @param  {Object}   performanceLog    performance data of performance run
 * @param  {String[]} metrics           asserted metrices
 * @param  {Function} [log=console.log] log method (for testing purposes)
 */
export const printResult = function (result, performanceLog, metrics, /* istanbul ignore next */ log = console.log) { // eslint-disable-line no-console
    log('\nPerformance Results\n===================')

    /**
     * filter out non performance metrics and sort displayed metrics based on
     * - asserted metrics first
     * - order of occurence (TTFB < load)
     */
    const resultsSorted = Object.entries(performanceLog.metrics)
        .filter(([metricName]) => PERFORMANCE_METRICS.includes(metricName))
        .sort((a, b) => {
            if (metrics.includes(a[0]) && !metrics.includes(b[0])) {
                return -1
            } else if (!metrics.includes(a[0]) && metrics.includes(b[0])) {
                return 1
            }

            return PERFORMANCE_METRICS.indexOf(a[0]) - PERFORMANCE_METRICS.indexOf(b[0])
        })

    for (const [metric, value] of resultsSorted) {
        const output = `${metric}: ${formatMetric[metric](value || 0)}`
        log(metrics.includes(metric) ? output : chalk.gray(output))
    }

    const resultDetails = []
    for (const [metric, { actual, lowerLimit, upperLimit }] of Object.entries(result.details)) {
        resultDetails.push(`Expected ${metric} to be between ${formatMetric[metric](lowerLimit)} and ${formatMetric[metric](upperLimit)} but was actually ${formatMetric[metric](actual)}`)
    }

    if (result.result === 'pass') {
        return log(`\nResult: ${result.result} ✅\n`)
    }

    return log(`\nPerformance assertions failed! ❌\n${resultDetails.join('\n')}\n`)
}

/**
 * wait for a specific condition to happen and return result
 */
export const waitFor = function (
    query,
    condition,
    errorMessage = 'job couldn\'t shutdown',
    interval = JOB_COMPLETED_INTERVAL,
    timeout = JOB_COMPLETED_TIMEOUT
) {
    if (typeof query !== 'function' || typeof condition !== 'function') {
        throw Error('Expect query and condition to by typeof function')
    }

    return new Promise((resolve, reject) => {
        const timeoutId = setTimeout(
            () => reject(new Error(errorMessage)),
            timeout)

        const intervalId = setInterval(async () => {
            const result = await query()
            if (!condition(result)) {
                return
            }

            clearTimeout(timeoutId)
            clearInterval(intervalId)
            return resolve(result)
        }, interval)
    })
}

/**
 * get list of metrics from job cli params and validate them
 * @param  {Object}   argv cli params
 * @return {String[]}      list of provided metrics if valid otherwise throws an error
 */
export const getMetricParams = function (argv) {
    const metrics = argv.all
        ? PERFORMANCE_METRICS
        : (argv.metric && !Array.isArray(argv.metric) ? [argv.metric] : argv.metric) || []

    /**
     * validate provided metrics
     */
    const invalidMetrics = metrics.filter((m) => !PERFORMANCE_METRICS.includes(m))
    if (invalidMetrics.length) {
        throw new Error(
            `You've provided invalid metrics: ${invalidMetrics.join(', ')}; ` +
            `only the following metrics are available: ${PERFORMANCE_METRICS.join(', ')}`)
    }

    return metrics
}

/**
 * get url to job details page of given test
 * @param  {Object} argv      cli params
 * @param  {String} sessionId of test
 * @return {String}           url to given test
 */
export const getJobUrl = function (argv, sessionId) {
    const hostname = (!argv.region || !argv.region.includes('eu') ? '' : 'eu-central-1.') + 'saucelabs.com'
    return `https://app.${hostname}/tests/${sessionId}`
}

export const formatMetric = {
    timeToFirstByte: prettyMs,
    firstPaint: prettyMs,
    firstContentfulPaint: prettyMs,
    firstMeaningfulPaint: prettyMs,
    domContentLoaded: prettyMs,
    timeToFirstInteractive: prettyMs,
    load: prettyMs,
    speedIndex: prettyMs,
    perceptualSpeedIndex: prettyMs,
    pageWeight: prettyBytes,
    pageWeightEncoded: prettyBytes
}

/**
 * print results of cli analyze command
 * @param  {Object}   jobResults        performance data
 * @param  {String[]} metrics           asserted metrices
 * @param  {Function} [log=console.log] log method (for testing purposes)
 */
export const analyzeReport = function (jobResults, metrics, /* istanbul ignore next */ log = console.log) { // eslint-disable-line no-console
    log('\nPerformance Results\n===================')

    if (Math.max(...jobResults.map((r) => r.results.length)) === 0) {
        return log('\n❌ FAILURE: your query did not seem to match any result!\n')
    }

    for (const result of jobResults) {
        log(`\n${result.passed ? '✅ SUCCESS:' : '❌ FAILURE:'} ${result.name}:`)

        const data = []
        data.push(['#', 'Url', 'Metrics'])

        for (const pageResult of result.results.sort((a, b) => a.orderIndex > b.orderIndex)) {
            const orderIndex = pageResult.orderIndex
            const url = pageResult.url
            const metricsOutput = Object.values(pageResult.metrics)
                .sort((a, b) => {
                    if (metrics.includes(a.metric) && !metrics.includes(b.metric)) {
                        return -1
                    } else if (!metrics.includes(a.metric) && metrics.includes(b.metric)) {
                        return 1
                    }

                    return PERFORMANCE_METRICS.indexOf(a.metric) - PERFORMANCE_METRICS.indexOf(b.metric)
                })
                .map(({ metric, value, baseline, passed }) => metrics.includes(metric)
                    ? `${chalk.bold(metric)}: ${formatMetric[metric](value)} ${passed
                        ? ''
                        : baseline.u < value
                            ? chalk.red(`(${formatMetric[metric](value - baseline.u)} over baseline)`)
                            : chalk.red(`(${formatMetric[metric](baseline.l - value)} under baseline)`)
                    }`
                    : chalk.gray(`${metric}: ${formatMetric[metric](value)}`)
                )
                .join('\n')

            data.push([orderIndex, url, metricsOutput])
        }

        log(table(data), `👀 Check out job at ${result.url}`)
    }

    log('')
}
