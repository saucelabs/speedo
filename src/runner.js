import { remote } from 'webdriverio'

import { getMetricParams, getThrottleNetworkParam } from './utils'
import ultradumbBenchmarkScript from './scripts/benchmark'
import userAgentScript from './scripts/userAgent'

const MAX_RETRIES = 3

/**
 * script that runs performance test on Sauce Labs
 *
 * @param  {String} username   name of user account
 * @param  {String} accessKey  access key of user account
 * @param  {Object} argv       command line arguments
 * @param  {String} name       name of the test
 * @param  {String} build      name of the build
 * @param  {String} logDir     path to directory to store logs
 * @return {Object}            containing result and detail information of performance test
 */
export default async function runPerformanceTest(username, accessKey, argv, name, build, logDir, retryCnt = 0, crmuxdriverVersion = 'stable') {
    const { site, platformName, browserVersion, tunnelIdentifier, parentTunnel } = argv
    const metrics = getMetricParams(argv)
    const networkCondition = getThrottleNetworkParam(argv)
    const sauceOptions = {
        'sauce:options': {
            name,
            build,
            crmuxdriverVersion,
            extendedDebugging: true,
            capturePerformance: true,
            seleniumVersion: '3.141.59',
            ...(tunnelIdentifier ? { tunnelIdentifier } : {}),
            ...(parentTunnel ? { parentTunnel } : {})
        }
    }

    const chromeOptions = {
        'goog:chromeOptions': { w3c: true }
    }

    const browser = await remote({
        user: username,
        key: accessKey,
        region: argv.region,
        logLevel: 'trace',
        outputDir: logDir,
        capabilities: {
            browserName: 'chrome',
            platformName,
            browserVersion,
            ...sauceOptions,
            ...chromeOptions
        }
    })

    const sessionId = browser.sessionId

    /**
     * throttle network
     */
    if (networkCondition !== 'online') {
        await browser.throttleNetwork(networkCondition)
    }

    /**
     * throttle CPU
     */
    if (argv.throttleCpu !== 0) {
        await browser.execute('sauce:debug', {
            method: 'Emulation.setCPUThrottlingRate',
            params: { rate: argv.throttleCpu }
        })
    }

    /**
     * open page
     *
     * if user is running with a Sauce Connect tunnel and tries to open
     * a localhost page we have to modify it to be 0.0.0.0 otherwise the
     * page won't be accessible
     */
    await browser.url(tunnelIdentifier
        ? site.replace('localhost', '0.0.0.0')
        : site
    )

    try {
        /**
         * run `assertPerformance` command within try/catch as command
         * can fail if performance command couldn't be captured for
         * any reasons, e.g. NO_NAVSTART
         */
        const result = await browser.assertPerformance(name, metrics)
        const benchmark = await browser.execute(ultradumbBenchmarkScript)
        const userAgent = await browser.execute(userAgentScript)

        await browser.deleteSession()
        return { sessionId, result, benchmark, userAgent }
    } catch (e) {
        await browser.deleteSession()

        /**
         * stop retrying after reaching MAX_RETRIES
         */
        if (retryCnt === MAX_RETRIES) {
            throw e
        }

        /**
         * log data couldn't be fetched due to a tracing issue
         * run test again:
         */
        return runPerformanceTest(username, accessKey, argv, name, build, logDir, ++retryCnt)
    }
}
