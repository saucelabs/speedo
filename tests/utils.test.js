import performanceResults from './__fixtures__/performance.json'
import {
    printResult, waitFor, getMetricParams, getThrottleNetworkParam,
    getJobUrl, analyzeReport, getJobName, getDeviceClassFromBenchmark
} from '../src/utils'
import { PERFORMANCE_METRICS } from '../src/constants'

const performanceLog = {
    timeToFirstByte: 123,
    firstContentfulPaint: 200,
    firstMeaningfulPaint: 222,
    load: 321,
    IdontexistMetric: 42,
    speedIndex: 10,
    pageWeight: 1000,
    score: 0.879935214
}

test('printResult when test passes', () => {
    const log = jest.fn()
    const result = { result: 'pass', details: {} }
    printResult(result, { metrics: performanceLog } , ['speedIndex', 'load', 'timeToFirstByte', 'score'], log)
    expect(log.mock.calls).toMatchSnapshot()
})

test('printResult', () => {
    const log = jest.fn()
    const result = {
        result: 'failed',
        details: { speedIndex: { actual: 10, lowerLimit: 3, upperLimit: 7 } }
    }

    printResult(result, { metrics: performanceLog } , ['speedIndex', 'load'], log)
    expect(log.mock.calls).toMatchSnapshot()
})

test('waitFor should throw if parameters are not functions', () => {
    expect(() => waitFor()).toThrow()
    expect(() => waitFor(jest.fn())).toThrow()
    expect(() => waitFor(null, jest.fn())).toThrow()
})

test('waitFor waits until condition is true', async () => {
    let i = false
    const query = async () => {
        await new Promise((resolve) => setTimeout(resolve, 10))
        return i
    }
    const condition = (_i) => {
        if (!_i) {
            i = true
            return _i
        }

        return _i
    }

    const start = Date.now()
    expect(await waitFor(query, condition, 10, 50)).toBe(true)
    expect(Date.now() - start).toBeGreaterThan(29)
})

test('waitFor times out if condition is never met', async () => {
    let i = false
    const query = async () => {
        await new Promise((resolve) => setTimeout(resolve, 100))
        return i
    }
    const condition = (_i) => {
        if (!_i) {
            i = true
            return _i
        }

        return _i
    }

    const customError = new Error('Error: uups')
    await expect(waitFor(query, condition, 'Error: uups', 10, 50))
        .rejects.toEqual(customError)
})

test('getMetricParams', () => {
    expect(getMetricParams({}))
        .toEqual([])
    expect(getMetricParams({ all: true }))
        .toEqual(PERFORMANCE_METRICS)
    expect(getMetricParams({ metric: 'load' }))
        .toEqual(['load'])
    expect(getMetricParams({ metric: ['load', 'speedIndex'] }))
        .toEqual(['load', 'speedIndex'])
    expect(() => getMetricParams({ metric: ['load', 'speedIndex', 'foobar'] }))
        .toThrow()
})

test('getThrottleNetworkParam', () => {
    expect(getThrottleNetworkParam({}))
        .toEqual('Good 3G')
    expect(getThrottleNetworkParam({ throttleNetwork: 'Regular 3G' }))
        .toEqual('Regular 3G')
    expect(() => getThrottleNetworkParam({ throttleNetwork: 'invalidNetworkCondition' }))
        .toThrow()
})

test('getJobUrl', () => {
    expect(getJobUrl({}, 'foobar'))
        .toEqual('https://app.saucelabs.com/performance/foobar/0')
    expect(getJobUrl({ region: 'eu' }, 'foobar'))
        .toEqual('https://app.eu-central-1.saucelabs.com/performance/foobar/0')
    expect(getJobUrl({ region: 'what?' }, 'foobar'))
        .toEqual('https://app.saucelabs.com/performance/foobar/0')
})

test('analyzeReport', () => {
    const log = jest.fn()
    analyzeReport(performanceResults , ['speedIndex', 'load', 'pageWeight'], log)
    expect(log.mock.calls).toMatchSnapshot()
})

test('getJobName', () => {
    expect(getJobName({ throttleCpu: 123 })).toBe('Performance test for undefined (on "Good 3G" and 123x CPU throttling)')
    expect(getJobName({ name: 'foobar' })).toBe('foobar')
})

test('getDeviceClassFromBenchmark', () => {
    expect(getDeviceClassFromBenchmark(1000)).toBe('desktop-class device')
    expect(getDeviceClassFromBenchmark(500)).toBe('high-end mobile phone')
    expect(getDeviceClassFromBenchmark(150)).toBe('mid-tier mobile phone')
    expect(getDeviceClassFromBenchmark(15)).toBe('budget mobile phone')
})
