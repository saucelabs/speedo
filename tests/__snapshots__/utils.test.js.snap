// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyzeReport 1`] = `
Array [
  Array [
    "
Performance Results
===================",
  ],
  Array [
    "
✅ SUCCESS: my performance test:",
  ],
  Array [
    "╔═══╤═════════════════════════╤══════════════════════════════╗
║ # │ Url                     │ Metrics                      ║
╟───┼─────────────────────────┼──────────────────────────────╢
║ 1 │ https://www.yahoo.com/  │ ✅ load: 9.8s                ║
║   │                         │ ✅ speedIndex: 1.3s          ║
║   │                         │ timeToFirstInteractive: 3.1s ║
║   │                         │ perceptualSpeedIndex: 2s     ║
║   │                         │ pageWeightEncoded: 8m 48.2s  ║
║   │                         │ timeToFirstByte: 502ms       ║
║   │                         │ domContentLoaded: 3.1s       ║
║   │                         │ firstPaint: 1.2s             ║
║   │                         │ firstContentfulPaint: 1.7s   ║
║   │                         │ firstMeaningfulPaint: 2s     ║
╟───┼─────────────────────────┼──────────────────────────────╢
║ 0 │ https://www.amazon.com/ │ ✅ load: 4.9s                ║
║   │                         │ ✅ speedIndex: 2s            ║
║   │                         │ timeToFirstInteractive: 3s   ║
║   │                         │ perceptualSpeedIndex: 2.2s   ║
║   │                         │ pageWeightEncoded: 9m 54.7s  ║
║   │                         │ timeToFirstByte: 693ms       ║
║   │                         │ domContentLoaded: 3s         ║
║   │                         │ firstPaint: 1.4s             ║
║   │                         │ firstContentfulPaint: 1.4s   ║
║   │                         │ firstMeaningfulPaint: 1.7s   ║
╚═══╧═════════════════════════╧══════════════════════════════╝
",
    "👀 Check out job at https://app.saucelabs.com/performance/83d526a5d20040a7b854a2cbef64c067/0
",
  ],
]
`;

exports[`getBudgetMetrics 1`] = `
Array [
  "estimatedInputLatency",
  "timeToFirstByte",
  "domContentLoaded",
  "firstVisualChange",
  "firstPaint",
  "firstContentfulPaint",
  "firstMeaningfulPaint",
  "lastVisualChange",
  "firstCPUIdle",
  "firstInteractive",
  "load",
  "speedIndex",
]
`;

exports[`getBudgetMetrics 2`] = `"You've provided invalid metrics in budget: foo; only the following metrics are available: estimatedInputLatency, timeToFirstByte, domContentLoaded, firstVisualChange, firstPaint, firstContentfulPaint, firstMeaningfulPaint, lastVisualChange, firstCPUIdle, firstInteractive, load, speedIndex, score, jankiness"`;

exports[`getJankinessParam 1`] = `
Object {
  "jankiness": Array [
    Object {
      "l": 50,
      "u": 100,
    },
  ],
}
`;

exports[`getJankinessParam 2`] = `
Object {
  "jankiness": Array [
    Object {
      "l": 50,
      "u": 100,
    },
  ],
}
`;

exports[`prepareBudgetData 1`] = `
Object {
  "domContentLoaded": Array [
    Object {
      "l": 200,
      "u": 300,
    },
  ],
  "estimatedInputLatency": Array [
    Object {
      "l": 0,
      "u": 20,
    },
  ],
  "firstCPUIdle": Array [
    Object {
      "l": 0,
      "u": 400,
    },
  ],
  "firstContentfulPaint": Array [
    Object {
      "l": 0,
      "u": 300,
    },
  ],
  "firstInteractive": Array [
    Object {
      "l": 0,
      "u": 400,
    },
  ],
  "firstMeaningfulPaint": Array [
    Object {
      "l": 0,
      "u": 300,
    },
  ],
  "firstPaint": Array [
    Object {
      "l": 0,
      "u": 300,
    },
  ],
  "firstVisualChange": Array [
    Object {
      "l": 0,
      "u": 100,
    },
  ],
  "lastVisualChange": Array [
    Object {
      "l": 0,
      "u": 400,
    },
  ],
  "load": Array [
    Object {
      "l": 50,
      "u": 500,
    },
  ],
  "speedIndex": Array [
    Object {
      "l": 100,
      "u": 500,
    },
  ],
  "timeToFirstByte": Array [
    Object {
      "l": 50,
      "u": 100,
    },
  ],
}
`;

exports[`printJankinessResult 1`] = `
Array [
  Array [
    "
Jankiness Results
===================",
  ],
  Array [
    "averageFPS: 60",
  ],
  Array [
    "idleTime: 5s",
  ],
  Array [
    "score: 98/100",
  ],
]
`;

exports[`printResult 1`] = `Array []`;

exports[`printResult when test passes 1`] = `Array []`;

exports[`sanitizeMetric 1`] = `"5s"`;

exports[`sanitizeMetric 2`] = `"50/100"`;

exports[`sanitizeMetric 3`] = `"80/100"`;

exports[`sanitizeMetric 4`] = `"10.03MB"`;

exports[`validateJankinessValue 1`] = `
Object {
  "jankiness": Array [
    Object {
      "l": 50,
      "u": 100,
    },
  ],
}
`;

exports[`validateJankinessValue 2`] = `
Object {
  "jankiness": Array [
    Object {
      "l": 50,
      "u": 100,
    },
  ],
}
`;

exports[`validateJankinessValue 3`] = `
Object {
  "jankiness": Array [
    Object {
      "l": 50,
      "u": 100,
    },
  ],
}
`;

exports[`validateJankinessValue 4`] = `"Jankiness score value should be between 0 to 100 you have provided 120"`;
